index.js

```diff
@@ -78,6 +83,196 @@ function loader (version) {
+      }
+
+      /**
+         *
+         * @param {Item} itemOne left hand item
+         * @param {Item} itemTwo right hand item
+         * @returns {[xpLevelCost, finalEnchs]}
+         * xpLevelCost is enchant data that is strictly from combining enchants
+         * finalEnchs is the array of enchants on the final object
+         */
+      function combineEnchants (itemOne, itemTwo, creative) {
+        const rightIsBook = itemTwo.name === 'enchanted_book'
+        const bothAreBooks = itemOne.name === 'enchanted_book' && rightIsBook
+        const findEnchantBy = (val, by) => Object.entries(mcData.enchantments).map(x => x[1]).find(x => x[by] === val)
+        const itemOneEnch = Item.getEnchants(itemOne)
+        const itemTwoEnch = Item.getEnchants(itemTwo)
+        if (itemOneEnch === null && itemTwoEnch === null) throw new Error('Both items are unenchanted')
+        const finalEnchs = []
+        let xpLevelCost = 0
+        for (const ench of itemTwoEnch) {
+          const enchOnItemOne = itemOneEnch.find(x => x.name === ench.name)
+          let { exclude, maxLevel, category, weight } = findEnchantBy(ench.name, 'name')
+          const { itemMultiplier, bookMultiplier } = getMultipliers(weight)
+          const categoryItems = armorMapping[category]
+          if (!bothAreBooks && !categoryItems.includes(itemOne.name) && !creative) continue
+          if (enchOnItemOne === undefined) { // first item doesn't have this ench
+            exclude = exclude.map(name => findEnchantBy(name, 'name'))
+            if (exclude.some(({ name }) => itemOneEnch.find(x => x.name === name))) { // has an excluded enchant
+              xpLevelCost++
+            } else {
+              const finalLevel = ench.lvl
+              xpLevelCost += rightIsBook ? (finalLevel * bookMultiplier) : (finalLevel * itemMultiplier)
+              finalEnchs.push({ id: ench.name, lvl: ench.lvl })
+            }
+          } else {
+            let finalLevel = 0
+            const itemOneLevel = enchOnItemOne.lvl
+            const itemTwoLevel = ench.lvl
+            if (itemOneLevel === itemTwoLevel) { // add check for max level > combined level
+              if (itemOneLevel === maxLevel) finalLevel = itemOneLevel
+              else finalLevel = itemOneLevel + 1
+              finalEnchs.push({ id: ench.name, lvl: finalLevel })
+              // return
+            } else if (itemTwoLevel > itemOneLevel) {
+              finalLevel = itemTwoLevel
+              finalEnchs.push({ id: ench.name, lvl: finalLevel })
+            }
+            xpLevelCost += rightIsBook ? (finalLevel * bookMultiplier) : (finalLevel * itemMultiplier)
+          }
+        }
+        for (const ench of itemOneEnch) {
+          if (!finalEnchs.find(x => x.name === ench.name)) {
+            finalEnchs.push(ench)
+          }
+        }
+        return { xpLevelCost, finalEnchs }
+      }
+      return combine(itemOne, itemTwo, creative, rename)
+    }
```

Comment:
All the functions inside Anvil should be declared outside, there is no need to re-declare them each time the function is called. Put them in Item if they need access to members, otherwise put them completely outside.

Reply:
done